### PR TITLE
Fix usage of size hints

### PR DESCRIPTION
**Pull request by bgamari**
This fixes a variety of size hint bugs in `text`'s fusion framework. It seems these were introduced when the `Size` type was extended to track both upper and lower bounds in f4fc30cea2a41bd26b6c4445843ba1b3f7d0dde8. These could manifest in a variety of issues similar to #197.

### CODE REVIEW

File: Data/Text/Internal/Fusion/Common.hs [hunk 1]
```diff
@@ -318,10 +318,13 @@ lengthI (Stream next s0 _len) = loop_length 0 s0
...
 compareLengthI :: Integral a => Stream Char -> a -> Ordering
-compareLengthI (Stream next s0 len) n =
-    case compareSize len (fromIntegral n) of
-      Just o  -> o
-      Nothing -> loop_cmp 0 s0
+compareLengthI (Stream next s0 len) n
+    -- Note that @len@ tracks code units whereas we want to compare the length
+    -- in code points. Specifically, a stream with hint @len@ may consist of
+    -- anywhere from @len/2@ to @len@ code points.
+  | Just LT <- compareSize (2*len) (fromIntegral n) = LT
```
Comment (Mistuke) [inline Data/Text/Internal/Fusion/Common.hs]: Is this one correct? Since `n` is in code points, and a code point can be 2 code units, won't
`len=6` and `n=1` give `LT` here while it should be `GT`? I would have thought that the original compare just lacked a call to `codePointsSize` over `n`?

File: Data/Text/Internal/Fusion/Common.hs [hunk 2]
```diff
@@ -699,7 +702,7 @@ replicateI n (Stream next0 s0 len) =
 -- (a,b), in which case, a is the next Char in the string, and b is
 -- the seed value for further production.
 unfoldr :: (a -> Maybe (Char,a)) -> a -> Stream Char
-unfoldr f s0 = Stream next s0 1 -- HINT maybe too low
+unfoldr f s0 = Stream next s0 unknownSize
```
Comment (Mistuke) [inline Data/Text/Internal/Fusion/Common.hs]: Don't you drop `s` in ever stream, so can't `unknownSize` be `charSize s`?

File: Data/Text/Internal/Fusion/Common.hs [hunk 3]
```diff
@@ -395,7 +398,7 @@ intersperse c (Stream next0 s0 len) = Stream next (I1 s0) len
 
 caseConvert :: (forall s. Char -> s -> Step (CC s) Char)
             -> Stream Char -> Stream Char
-caseConvert remap (Stream next0 s0 len) = Stream next (CC s0 '\0' '\0') len
+caseConvert remap (Stream next0 s0 _) = Stream next (CC s0 '\0' '\0') unknownSize
```
Comment (Mistuke) [inline Data/Text/Internal/Fusion/Common.hs]: hmm unless I'm misunderstanding something, can't this be `len + maxSize $ charSize s0` ?

File: Data/Text/Internal/Fusion/Common.hs (hunk 2 above)
Comment (bgamari) [reply Data/Text/Internal/Fusion/Common.hs]: I'm not sure I follow. Can you rephrase?

File: Data/Text/Internal/Fusion/Common.hs (hunk 1 above)
Comment (bgamari) [reply Data/Text/Internal/Fusion/Common.hs]: Oh dear, yes, I believe you are right.

File: Data/Text/Internal/Fusion/Common.hs (hunk 2 above)
Comment (Mistuke) [reply Data/Text/Internal/Fusion/Common.hs]: What I mean is, in each step of the stream you consume one input character don't you? So according to the comment

     -- The function takes the element and returns
     -- Nothing if it is done producing the stream or returns Just
     -- (a,b), in which case, a is the next Char in the string, and b is 
    -- the seed value for further production. 

So isn't the initial length the number of code points of the seed?

Comment (bgamari) [reply Data/Text/Internal/Fusion/Common.hs]: Ahh, I see.
> So isn't the initial length the number of code points of the seed?

Not quite, since the user might take and double the size of each code point (turning one-unit code-point into a two-unit code point). Consequently, I think we can set the upper bound at `2*hint`.

Comment (bgamari) [reply Data/Text/Internal/Fusion/Common.hs]: Let's try this again.

> What I mean is, in each step of the stream you consume one input character don't you?

This is where you lose me. What input stream are we talking about here? `unfoldr` produces a stream by unfolding starting from an initial value, successively applying the given function until it returns `Nothing`. Consequently I'm not sure how we could possibly know how long the stream will be; it depends entirely on the function.

File: Data/Text/Internal/Fusion/Common.hs (hunk 3 above)
Comment (bgamari) [reply Data/Text/Internal/Fusion/Common.hs]: Okay, so I looked at the call sites and indeed they *do* produce only one code point per input code point although there is of course nothing in the type of `caseConvert` that requires this be the case. However, given that `caseConvert` isn't exported, I suppose we can just state this as an invariant.

Comment (bgamari) [reply Data/Text/Internal/Fusion/Common.hs]: On more careful reading it seems that the point I made above actually wasn't quite right: `remap` can produce up to three codepoints for every input code-point.

Comment (Mistuke) [reply Data/Text/Internal/Fusion/Common.hs]: Ah ok, thanks for the explanation!

File: Data/Text/Internal/Fusion/Common.hs (hunk 2 above)
Comment (Mistuke) [reply Data/Text/Internal/Fusion/Common.hs]: Ah, there's my misunderstanding. I had thought it only applied it once to get the next value. In which case I figured you always know the max amount of code points the next would produce. Thanks for clearing that up!

**Comment by hvr** [comment]
Looks good to me; Thanks everyone!

**Comment by bgamari** [comment]
Can we get a release with this soon?
